### PR TITLE
Replace TableDataPrefix with TableData{Min,Max}.

### PR DESCRIPTION
**Pull request by petermattis**
This saves 1-byte per key encoded at the SQL layer, but it makes usage of
non-SQL keys in tests a bit trickier. Most tests were "fixed" by prepending
"\xff" to their keys. For easier readability, the cli tests were fixed by
prepending "~".

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3376)

### CODE REVIEW

**Comment by petermattis** [comment]
A better option might be to rearrange the encoding tags so that the `[IntMin,IntMax]` range falls outside of the ascii character set.

**Comment by tschottdorf** [comment]
Definitely doesn't make it easier, more readable or more fun to write correct KV tests. Is rearranging the tags iffy? Otherwise, perhaps some tooling (`testutils.{R,}Key("x") == "\xff\x00x`" where `\xff\x00` is stripped during pretty printing? It should be clear when looking at the tests why they consistently chose to use weird keys so that writers of new tests know what's going on.

**Comment by petermattis** [comment]
Added another commit which did as I suggested. The int tag range was moved to `[0x80,0xff]`. Now it is safe to use ascii characters as the prefix for keys in tests. I also took the opportunity to expand the small int range to values `<=111`. I think I was mistaken about my earlier concern about not being able to decode `VarintDecreasing` keys for pretty-printing. We'll be able to decode such keys, but they will simply look like the inverse of what was encoded which seems acceptable.

File: sql/table.go [hunk 1]
```diff
@@ -393,11 +392,10 @@ func makeKeyVals(desc *TableDescriptor, columnIDs []ColumnID) ([]parser.Datum, e
 }
 
 func decodeIndexKeyPrefix(desc *TableDescriptor, key []byte) (IndexID, []byte, error) {
-	if !bytes.HasPrefix(key, keys.TableDataPrefix) {
+	if encoding.PeekType(key) != encoding.Int {
 		return 0, nil, util.Errorf("%s: invalid key prefix: %q", desc.Name, key)
```
Comment (tamird) [inline sql/table.go]: this message is less helpful now

File: sql/scan.go [hunk 2]
```diff
@@ -74,11 +73,9 @@ type span struct {
 // prettyKey pretty-prints the specified key, skipping over the first skip
 // fields.
 func prettyKey(key roachpb.Key, skip int) string {
-	if !bytes.HasPrefix(key, keys.TableDataPrefix) {
-		return fmt.Sprintf("index key missing table data prefix: %q vs %q",
-			key, keys.TableDataPrefix)
+	if encoding.PeekType(key) != encoding.Int {
+		return fmt.Sprintf("index key missing table data prefix: %q", key)
```
Comment (tamird) [inline sql/scan.go]: this message is less helpful now

File: sql/table.go (hunk 1 above)
Comment (petermattis) [inline sql/table.go]: You'll still be able to see what the key prefix is and that it isn't a valid int tag. Do you have a suggestion for something better?

File: sql/scan.go (hunk 2 above)
Comment (petermattis) [inline sql/scan.go]: Adjusted the message to not mention "table data", but I'm not sure what else you want. I want to see the key from which I'll be able to see that the first byte is not a valid int tag.

File: keys/constants.go [hunk 3]
```diff
@@ -160,13 +160,10 @@ var (
...
 
-	// TableDataPrefix prefixes all table data. It is specifically chosen to
-	// occur after the range of common user data prefixes so that tests which use
-	// those prefixes will not see table data.
-	//
-	// TODO(peter): Replace with TableDataMin and TableDataMax which are the min
-	// and max int tag values from util/encoding/encoding.go:int{Min,Max}.
-	TableDataPrefix = roachpb.Key("\xff")
+	// TableDataMin is the start of the range of table data keys.
+	TableDataMin = roachpb.Key{0x80}
```
Comment (bdarnell) [inline keys/constants.go]: This should reference `encoding.IntMin` and `IntMax`.

File: storage/engine/mvcc_test.go [hunk 4]
```diff
@@ -1923,7 +1923,7 @@ func TestFindSplitKey(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	ind, _ := strconv.Atoi(string(humanSplitKey))
+	ind, _ := strconv.Atoi(string(humanSplitKey[1:]))
```
Comment (bdarnell) [inline storage/engine/mvcc_test.go]: It looks like this is no longer necessary.

File: storage/engine/mvcc_test.go [hunk 5]
```diff
@@ -2099,7 +2099,7 @@ func TestFindBalancedSplitKeys(t *testing.T) {
 		ms := &MVCCStats{}
 		var expKey roachpb.Key
 		for j, keySize := range test.keySizes {
-			key := roachpb.Key(fmt.Sprintf("%d%s", j, strings.Repeat("X", keySize)))
+			key := roachpb.Key(fmt.Sprintf("\xff%d%s", j, strings.Repeat("X", keySize)))
```
Comment (bdarnell) [inline storage/engine/mvcc_test.go]: Still needed?

File: util/encoding/encoding.go [hunk 6]
```diff
@@ -50,6 +45,13 @@ const (
 
 	bytesMarker byte = floatInfinity + 1
 	timeMarker  byte = bytesMarker + 1
+
+	// intMin is chosen such that the range of int tags does not overlap the
+	// ascii character set that is frequently used in testing.
+	intMin   = 0x80
+	intZero  = intMin + 8
+	intSmall = 111
```
Comment (bdarnell) [inline util/encoding/encoding.go]: Where did this come from? Is it the value that makes `intMax` equal to `0xff`? If so maybe they should go the other way:
```
intMax = 0xff
intSmall = intMax - intZero - 8
```

**Comment by bdarnell** [comment]
LGTM

File: keys/constants.go (hunk 3 above)
Comment (petermattis) [inline keys/constants.go]: Done.

File: storage/engine/mvcc_test.go (hunk 5 above)
Comment (petermattis) [inline storage/engine/mvcc_test.go]: Nope, just missed it.

File: util/encoding/encoding.go (hunk 6 above)
Comment (petermattis) [inline util/encoding/encoding.go]: Good point. Changed.

**Pull request merged**